### DISCUSSION TITLE
Tailwind Base CSS Variable Performance

### DISCUSSION THREAD

**Discussion by samouri** [tailwindlabs/tailwindcss#14506] [General]
**What version of Tailwind CSS are you using?**

For example: v3.4.13

**What build tool (or framework if it abstracts the build tool) are you using?**

N/A

**What version of Node.js are you using?**

N/A

**What browser are you using?**

Latest Chrome

**What operating system are you using?**

Windows

**Reproduction URL**

https://play.tailwindcss.com/, no modifications needed

**Describe your issue**

👋 Hi all.

First off I'd like to thank y'all for tailwind. I really love using it.

While diagnosing performance problems on a closed-source app I develop, I realized that `@base` is adding a host of css variables to the `*` selector. I'm confused why it's targeted this way instead of [:root](https://developer.mozilla.org/en-US/docs/Web/CSS/:root#examples).

I *think* targeting `*` is an issue for css variables in particular because it appears to redefine the variable at every single node. That's why when inspecting any element on the page and looking at the variable definitions, you'll often see dozens of overriden definitions. Not only is this noisy, but I've measured a significant performance impact to `Recalculate Styles` in Chrome Devtools.

*Performance impact details*

It's likely a different level of impact for every app, but I'm measuring around a ~2.5x hit to Recalculate Styles on Tailwind Play. The closed-source app I develop involves a page with a supidly large number of DOM elements and has a larger impact. It likely scales with the depth of nodes. See the attached screenshots for traces I took on Tailwind Play with the css vars applied to `*` vs. `:root`.

| * selector  |  :root selector |
| ------------- | ------------- |
| 5.18ms  | 1.96ms |

Repro steps: do the below steps for both default and after modifying the css var initialization selector to `:root`.
1. Go to https://play.tailwindcss.com/.
2. Click the "Preview only" button.
3. Now start a performance trace and click the "Vertical Split" option.
4. Stop the trace and examine "Recalculate Styles".

*Applied to star (current behavior)*
<img width="1919" alt="applied-to-star" src="https://github.com/user-attachments/assets/03805a34-b6cf-4772-8930-7b2022016bf1">

*Applied to root*
<img width="1920" alt="applied-to-root" src="https://github.com/user-attachments/assets/d52e1255-a373-44b4-aa79-aa9aae2906e9">

*Request for advice*

I'm considering manually injecting the required css variables to `:root` to solve these issues instead of using `@base`.
Do y'all think that's an ok workaround? Are there things I should be wary of?

**Comment by adamwathan** [comment]
Hey! Check out this previous discussion here: https://github.com/tailwindlabs/tailwindcss/discussions/13828

TL;DR is the behavior is very different if we set them on `:root` and it'll break tons of stuff 😄

In v4 we're able to use `@property` instead which is an exciting improvement, but you can also try the `optimizeUniversalDefaults` option in v3 to improve things a bit. We tried to make that the default but its' incompatible with CSS modules and Vue style blocks, so unfortunately not safe to make default.

**Comment by samouri** [comment]
Thank you for the quick response @adamwathan!
Your article on composing the uncomposable was super informative and explains the motivation for the current behavior.

Am I right in thinking the `optimizeUniversalDefaults` option is still experimental?
I see it in the typescript types for `Config`, but is there also any documentation for it?

Even though I don't use CSS modules or Vue style blocks, I'm leaning against enabling this in production due to the experimental label.

**Reply by adamwathan** [reply]
It's very robust as long as you aren't using CSS modules or Vue/Svelte style blocks 👍  We have it enabled for tailwindcss.com for what it's worth.

<img width="456" alt="image" src="https://github.com/user-attachments/assets/94a64ea5-f358-44d3-a42d-e476393b1b81">
